### ISSUE TITLE
VICI doesn't use default ID if ID is not confirmed by certificate

### ISSUE THREAD

**Issue by zendulkaj** [strongswan/strongswan#862]
**System (please complete the following information):**
 - OS: Embedded Kernel Linux
 - Kernel version (if applicable):  4.14.138
 - strongSwan version(s):  5.9.5
 - Tested/confirmed with the latest version: yes

**Describe the bug**
If wrong ID (ID is not confimed by certificate) is configured then stroke_config.c uses defualt ID. 
https://git.strongswan.org/?p=strongswan.git;a=blob;f=src/libcharon/plugins/stroke/stroke_config.c;h=175b6b5495e29d80312a941dd70a408f2b8f1d38;hb=HEAD#l442

vici_config.c does not verify ID so then it is not possible to establish IPsec connection with wrong ID. This different behaviour (stroke vs. vici) is a issue when remote device (with wrong configured ID) is updated because after update the device is not able to connect via IPsec anymore. 

**To Reproduce**
1. Set ID which is not defined in SAN of certificate.

**Expected behavior**
If ID is not confirmed by certificate then it should be used the defualt ID ( =Common name)

**Logs/Backtraces**
2022-01-24 07:47:30 charon: 12[CFG] loaded certificate 'C=DE, ST=sn, O=l, OU=it, CN=testrouter01' 
2022-01-24 07:47:30 charon: 12[CFG] loaded certificate 'C=DE, ST=sn, O=l, OU=it, CN=ipfire.localdomain' 
2022-01-24 07:47:30 charon: 12[CFG] loaded certificate 'C=DE, ST=sn, L=bach, O=l, OU=it, CN=l CA, E=ad@l.de' 
2022-01-24 07:47:30 charon: 15[CFG] loaded ANY private key 
2022-01-24 07:47:31 charon: 15[CFG] added vici connection: ipsec1 
2022-01-24 07:47:31 charon: 15[CFG] initiating 'ipsec1' 
2022-01-24 07:47:31 charon: 15[IKE] initiating Main Mode IKE_SA ipsec1[1] to 192.168.101.46 
2022-01-24 07:47:31 charon: 15[IKE] no private key found for '192.168.3.1' 
2022-01-24 07:47:31 charon: 15[CFG] configuration uses unsupported authentication

**Comment by tpaukrt** [comment]
Proposed patch for this issue:
```
diff --exclude CVS --exclude .git -uNr strongswan-5.9.5/src/libcharon/plugins/vici/vici_config.c strongswan-5.9.5.modified/src/libcharon/plugins/vici/vici_config.c
--- strongswan-5.9.5/src/libcharon/plugins/vici/vici_config.c	2022-01-08 12:54:02.000000000 +0100
+++ strongswan-5.9.5.modified/src/libcharon/plugins/vici/vici_config.c	2022-01-25 15:30:16.902373551 +0100
@@ -2083,6 +2083,8 @@
 		pubkey_cert_t *pubkey_cert;
 		identification_t *id;
 		bool default_id = FALSE;
+		bool has_subject = FALSE;
+		certificate_t *first = NULL;
 
 		INIT(auth,
 			.request = peer->request,
@@ -2109,6 +2111,18 @@
 					auth->cfg->add(auth->cfg, AUTH_RULE_IDENTITY, id->clone(id));
 					default_id = TRUE;
 				}
+				else if (cert->get_type(cert) != CERT_TRUSTED_PUBKEY &&
+						 id->get_type != ID_ANY)
+				{
+					if (!first)
+					{
+						first = cert;
+					}
+					if (cert->has_subject(cert, id))
+					{
+						has_subject = TRUE;
+					}
+				}
 				else if (cert->get_type(cert) == CERT_TRUSTED_PUBKEY &&
 						 id->get_type(id) != ID_ANY)
 				{
@@ -2120,6 +2134,14 @@
 		}
 		enumerator->destroy(enumerator);
 
+		if (first && !has_subject)
+		{
+			DBG1(DBG_CFG, "  id '%Y' not confirmed by certificate, "
+					"defaulting to '%Y'", id, first->get_subject(first));
+			id = first->get_subject(first);
+			auth->cfg->add(auth->cfg, AUTH_RULE_IDENTITY, id->clone(id));
+		}
+
 		auths = strcasepfx(name, "local") ? peer->local : peer->remote;
 		enumerator = auths->create_enumerator(auths);
 		while (enumerator->enumerate(enumerator, &current)) 
```

**Comment by tobiasbrunner** [comment]
> vici_config.c does not verify ID so then it is not possible to establish IPsec connection with wrong ID.

Which produces an error and you can correct the config (e.g. delete the identity so it defaults to the certificate's subject DN), instead of using an identity that was clearly not intended to be used (why would an identity be configured otherwise?), which might cause issues on the other peer if it doesn't expect the full subject DN.

> This different behaviour (stroke vs. vici) is a issue when remote device (with wrong configured ID) is updated because after update the device is not able to connect via IPsec anymore.

How did you convert the configs? Why would you not check the config during that process (or just delete the local identity if you want to use the subject DN as identity)?

**Comment by zendulkaj** [comment]
I fully understand that it is not correct configured ID. We suggested to customers to delete wrong configured IDs but many customers configured wrong IDs in the past on many devices (hundreds) and ipsec.conf (stroke) established ipsec tunnel with this ID misconfiguration and reported only a notice "ID is not confirmed by certificate, defaulting to ...". So customers stayed with wrong IDs - nothing enforced them to fix ipsec configuration. The issue is that customers found wrong ID misconfiguration  on many remote devices. After upgrading FW they would lost remote access to their devices. It is very very uncomfortable and time consumig for customers to check every device if ID is correct or not. Especially when it worked with older FW (ipsec.conf). We are forced to use that patch as customers are urgently asking us for backward compatible FW.

**Comment by tobiasbrunner** [comment]
> So customers stayed with wrong IDs - nothing enforced them to fix ipsec configuration.

That's exactly the point. The config was wrong in the first place, why would we encourage users to continue to do this?

> After upgrading FW they would lost remote access to their devices.

How does a firmware upgrade cause a switch from ipsec.conf to swanctl.conf/vici without config changes?  (Manual or automated - in either case it should be possible to fix/remove the wrong identities.)

**Comment by zendulkaj** [comment]
Generally it is not possible for us to use any scripts or someone like that to correct IDs during update FW as it is not known if ID is correct or not. We would have to verify certificate if match with ID or not what we find very complicated. It is only up to customers to correct wrong IDs and this the issue - they do not understand why it worked before and now not. From their point of view it is bug. Anyway I understand you - the config is wrong. 

Do you find anything problematic (concerning the code placement, etc.) about that patch? Thanks

**Comment by tobiasbrunner** [comment]
> We would have to verify certificate if match with ID or not what we find very complicated.

Yes, you'd have to check it against the SANs and otherwise remove it to default to the subject DN. It's possible to do this with command line tools. For instance, to check if `dave@strongswan.org` is contained as a SAN in a certificate something like this works ([source](https://superuser.com/a/1275155/98749)):

```
for san in $(openssl x509 -in testing/hosts/dave/etc/swanctl/x509/daveCert.pem -noout -text | grep -A1 'Subject Alternative Name' | tail -n1 | tr -d ','); { echo $san | cut -f 2 -d ':'; } | grep dave@strongswan.org
```
This doesn't verify the type, but you could do that by removing `cut` and matching with `DNS:` or `email:` prefix (if IP addresses are used as SANs, it also matches the given ID against `IP` and the prefix would be `Address:` because `IP Address:` is split up).

Checking identities against certificates might be a feature that could be added to `pki --verify`, which would benefit from using strongSwan's actual [identity parsing](https://wiki.strongswan.org/projects/strongswan/wiki/IdentityParsing) and matching code.

> Do you find anything problematic (concerning the code placement, etc.) about that patch? Thanks

It does have the same bug the pubkey code had, which i fixed recently (e43052893dd7464572e1533767fdc7f6a42ea4ee). Other than that it looks OK.

**Comment by zendulkaj** [comment]
Many thanks for the advice and review the patch.
